### PR TITLE
 C++ sensor/actuator interfaces

### PR DESCRIPTION
**Pull request by whbruce**
Added C++ interfaces for the following sensor/actuator types
IADC
ICO2Sensor
IHumiditySensor
ILightController
ILightSensor
IModuleStatus
IPressureSensor
ITemperatureSensor

Added following modules which implement interfaces as shown
BMP180:   IPressureSensor, ITemperatureSensor
BMP280:   IPressureSensor, ITemperatureSensor, IHumiditySensor
DS1808LC: ILightController
LP8860:   ILightController
MAX44009: ILightSensor
SI1132:   ILightSensor
SI7005:   ITemperatureSensor, IHumiditySensor
T6713:    ICO2Sensor

Added interface implementation to the following existing modules
ADS1X15:  IADC

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -81,7 +90,7 @@ See building documentation [here](docs/building.md).
 
 Porting [link](docs/porting.md) has more information on making new UPM modules.
 
-There is also an example available for max31855 [sensor](docs/max31855.md).
+There is also an example available gfor max31855 [sensor](docs/max31855.md).
```
Comment (pylbert) [inline README.md]: Small typo

File: src/upm/iLightSensor.h [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+ * @brief ILightSensor Interface for Light Sensors
+ */
+ 
+/**
+ *
+ * @brief Interface for Light Sensors
+ 
+ * This interface is used to represent light sensors
+
+ * @snippet light-sensor.cxx Interesting
```
Comment (pylbert) [inline src/upm/iLightSensor.h]: Possible small typo

File: src/upm/iModuleStatus.h [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+#include <stdexcept>
+
+namespace upm
+{
+/**
+ * @brief Interface for Module Status. Sensor and Actuactor Interfaces Derive from this Interface.
+ */
+
+#define UPM_THROW(msg) throw std::runtime_error(std::string(__FUNCTION__) + ": " + (msg))
```
Comment (pylbert) [inline src/upm/iModuleStatus.h]: Possibly add __LINE__ and some flavor of __PRETTY_FUNCTION__ 
i.e.   /path/to/filename:namespace:class:function:linenumber

File: src/upm/iModuleStatus.h [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    * @return true if correctly configured, false it not
+    */
+   virtual bool isConfigured() = 0;
+
+   /**
+    * Returns name of module. This is the string in library name after libupm_
+
+    * @return name of module
+    */
+    virtual const char* getModuleName() = 0;
```
Comment (pylbert) [inline src/upm/iModuleStatus.h]: Should this be a non-pointer return type?  This leaves it up to the implementing class to create a valid chunk of memory to return which may not be handled correctly by calling code.

Comment (whbruce) [inline src/upm/iModuleStatus.h]: I chose to return `const char*` as STL does not define an ABI so exporting objects such as `std::string` from a shared library can be dangerous. However the mraa C++ API exports `std::string` so I should probably do likewise to be consistent.

**Comment by Propanu** [comment]
All merged, thanks again for this great contribution!

**Pull request closed without merging**